File: tests/fixtures/ocaml/patterns.ml
let case_scope input =
  let result =
    match input with
    | Some callback when callback input -> callback input
    | None -> fallback input
  in
  callback result

let pair_scope input =
  let (left, right) = split input in
  left right

let function_scope input =
  let apply = function
    | Some callback -> callback input
    | None -> fallback input
  in
  apply input

let exception_scope action =
  try action () with
  | Failure callback -> callback ()

File: tests/fixtures/haskell/patterns.hs
caseScope input =
  let result = case input of
        Just callback -> callback input
        Nothing -> fallback input
  in callback result

pairScope input =
  let (left, right) = split input
  in left right

doScope input = do
  (action, value) <- fetch input
  action value

comprehensionScope values =
  [action value | (action, value) <- values, keep value]
